### PR TITLE
Add proptest generation

### PR DESCRIPTION
**Pull request by lmondada**
A simple function to generate random (non-empty) portgraphs for `proptest`.

### CODE REVIEW

File: src/proptest.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    ///  - `max_port` is the maximum number of incoming and outgoing ports at
+    ///    every node
+    fn gen_no_edge_graph(max_n_nodes: usize, max_port: usize)(
+        ports in prop::collection::vec(0..=max_port, 2..=max_n_nodes)
+    ) -> PortGraph {
+        let mut g = PortGraph::new();
+        let mut ind = 0;
+        while ind + 1 < ports.len() {
+            g.add_node(ports[ind], ports[ind + 1]);
+            ind += 2;
```
Comment (aborgna-q) [inline src/proptest.rs:22]: Doesn't this generate `max_n_nodes/2` nodes at max? Perhaps generating `(# inputs, # outputs)` tuples for `ports`, and having a `for` loop here would be clearer.

File: src/proptest.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        n_edges = [n_edges, in_ports.len(), out_ports.len()]
+            .into_iter()
+            .min()
+            .unwrap();
+        if in_ports.len() > n_edges {
+            in_ports.drain(n_edges..);
+        }
+        if out_ports.len() > n_edges {
+            out_ports.drain(n_edges..);
+        }
```
Comment (aborgna-q) [inline src/proptest.rs:61]:
```suggestion
        in_ports.truncate(n_edges);
        out_ports.truncate(n_edges);
```

File: src/proptest.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    let graph = gen_no_edge_graph(max_n_nodes, max_port);
+    (0..=max_n_edges, graph).prop_perturb(|(mut n_edges, graph), mut rng| {
+        let mut in_ports = Vec::new();
+        let mut out_ports = Vec::new();
+        for p in graph.ports_iter() {
+            match graph.port_direction(p).unwrap() {
+                Direction::Incoming => in_ports.push(p),
+                Direction::Outgoing => out_ports.push(p),
+            }
+        }
```
Comment (aborgna-q) [inline src/proptest.rs:48]: Just nitpicking, but check `Iterator::partition` next time :)

**Review by aborgna-q — CHANGES_REQUESTED** [review]
Some comments, but LGTM.

Do you know how this plays with shrinking? The (nodes, ports, edges) tuple tries to simplify the fields in order, but maybe we'd want to reverse it, since reducing the edges is the only thing that doesn't alter all the graph structure?

File: src/proptest.rs (hunk 3 above)
Comment (lmondada) [reply src/proptest.rs:48]: Oh excellent, super useful! Thank you for pointing that out, please keep teaching me things :)

File: src/proptest.rs (hunk 1 above)
Comment (lmondada) [reply src/proptest.rs:22]: Yep, you're right. Fixed.

File: src/proptest.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
+use std::iter::zip;
```
[the remark this answers is not in the document]
Comment (lmondada) [reply src/proptest.rs:5]: Done.

**Comment by lmondada** [comment]
> Do you know how this plays with shrinking?

It's my first time building `proptest` strategies, so I'm not familiar with the details of the shrinking ordering. I'll happily take your word for it: I've swapped the ordering of the tuple `gen_graph_and_edges`, do you think this should be better?

Thanks for your comments, super useful!

**Pull request merged**